libvast/vast/fbs/table.hpp

```diff
@@ -0,0 +1,171 @@
+    }
+  }
+
+  /// Default copy-construction and copy-assignment.
+  table(const table& other) = default;
+  table& operator=(const table& rhs) = default;
+
+  /// Default move-construction and move-asignment.
+  table(table&& other) noexcept = default;
+  table& operator=(table&& rhs) noexcept = default;
+
+  /// Default destructor.
+  virtual ~table() noexcept = default;
+
+  // -- concepts ---------------------------------------------------------------
+
+  /// Return a view on the underlying byte buffer.
+  /// @pre `x`
+  friend span<const byte> as_bytes(const derived_type& x) noexcept {
+    VAST_ASSERT(x);
+    return as_bytes(x.chunk_);
+  }
+
+  /// Serialize/deserialize directly from/to the underlying byte buffer.
+  template <class Inspector>
+  friend auto inspect(Inspector& f, derived_type& x) ->
+    typename Inspector::result_type {
+    return f(caf::meta::type_name(table_name()), x.chunk_,
+             caf::meta::load_callback([&] { return x.load(); }),
+             caf::meta::save_callback([&] { return x.save(); }));
+  }
+
+  // -- properties -------------------------------------------------------------
+
+  /// Access the underlying FlatBuffers root table.
+  /// @pre `*this`
+  const root_type& operator*() const noexcept {
+    return *root();
+  }
+
+  /// Access the underlying FlatBuffers root table.
+  /// @pre `*this`
+  const root_type* operator->() const noexcept {
+    return root();
+  }
+
+  /// Check whether the FlatBuffers table is valid.
+  explicit operator bool() const noexcept {
+    return chunk_ != nullptr;
+  }
+
+  /// Returns the size of the underlying chunk in Bytes.
+  size_t num_bytes() const noexcept {
+    return *this ? chunk_->size() : 0;
+  }
+
+protected:
+  // -- opt-in properties ------------------------------------------------------
+
+  /// Queries whether there is exactly once reference.
+  bool unique() const noexcept {
+    return chunk_->unique();
+  }
+
+  /// Adds an additional step for deleting this table.
+  /// @param f Function object that gets called after all previous deletion
+  /// steps ran.
+  template <class F>
+  void add_deletion_step(F&& f) const noexcept {
+    chunk_->add_deletion_step(std::forward<F>(f));
+  }
+
+  /// Access the underlying chunk.
+  const chunk_ptr& chunk() const noexcept {
+    return chunk_;
+  }
+
+  /// Customization point for loading after deserialization of the chunk.
+  /// @note The default implementation does nothing.
+  virtual caf::error load() {
+    return caf::none;
+  }
+
+  /// Customization point for saving after serialization of the chunk.
+  /// @note The default implementation does nothing.
+  virtual caf::error save() {
+    return caf::none;
+  }
+
+private:
+  // -- implementation details -------------------------------------------------
+
+  /// Access the underlying FlatBuffers root table.
+  /// @pre `*this`
+  /// @post `result != nullptr`
+  const root_type* root() const noexcept {
+    VAST_ASSERT(*this);
```

Comment:
`VAST_ASSERT(this)`

Reply:
That's wrong. We want to check whether the `chunk_` is set, not the `this` pointer of the wrapping table.

Reply:
Ah, right. `VAST_ASSERT(chunk_)` might be a bit less confusing.

Reply:
I intentionally used the assertion as is, because it matches the pre-condition. The chunk may not be exposed to the user.